### ISSUE TITLE
lt-nghttpd linked against wrong libssl

### ISSUE THREAD

**Issue by bekoli** [tatsuhiro-t/nghttp2#215]
tl;dr: Setting _PKG_CONFIG_PATH_ for another libssl is not enough, you need _LD_LIBRARY_PATH_ as well or _ld_ will use the system library.

In detail:

I compiled libssl 1.0.2a for ALPN testing and installed it to `/usr/local` to not override my system libssl:

> beko@ernie ~ $ cat /usr/local/lib/pkgconfig/openssl.pc  
> prefix=/usr/local
> exec_prefix=${prefix}
> libdir=${exec_prefix}/lib
> includedir=${prefix}/include
> 
> Name: OpenSSL
> Description: Secure Sockets Layer and cryptography libraries and tools
> Version: 1.0.2a
> Requires: libssl libcrypto

I set PKG_CONFIG_PATH according to `./configure --help`

> beko@ernie ~ $ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/
> beko@ernie ~ $ echo $PKG_CONFIG_PATH
> /usr/local/lib/pkgconfig/

Now I enter my build directory, run configure and stuff. Checking the Makefile I get:

> beko@ernie ~/Downloads/nghttp2-0.7.12 $ grep ^OPENSSL Makefile
> OPENSSL_CFLAGS = -I/usr/local/include
> OPENSSL_LIBS = -L/usr/local/lib -lssl -lcrypto

All looks good. After the make I run `./src/nghttpd` and end up with  `/lib/x86_64-linux-gnu/libssl.so.1.0.0` in the now (on runtime, not maketime) created `lt-nghttpd` however:

> beko@ernie ~/Downloads/nghttp2-0.7.12 $ ldd src/.libs/lt-nghttpd | grep libssl
>	libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f7a1b1b3000)

This results in a symbol lookup error:

> nghttp2-0.7.12/src/.libs/lt-nghttpd: undefined symbol: SSL_CTX_set_alpn_select_cb

I can persuade this to run by making use of LD_PRELOAD:

> LD_PRELOAD=/usr/local/lib/libssl.so:/usr/local/lib/libcrypto.so ./src/nghttpd 8080 ../my.key ../my.pem

Checking `./src/nghttpd` I find a _relink_command_ where `-L/usr/local/lib -lssl -lcrypto`  is set so it's an order issue (system libs are grabbed first) and indeed the linking is done correctly when I set LD_LIBRARY_PATH:

> beko@ernie ~/Downloads/nghttp2-0.7.12 $ export LD_LIBRARY_PATH=/usr/local/lib/
> beko@ernie ~/Downloads/nghttp2-0.7.12 $ `./src/nghttpd`

Now I get the lib linked to the version I wanted to have in the first place:

> beko@ernie ~/Downloads/nghttp2-0.7.12 $  ldd src/.libs/lt-nghttpd | grep libssl
>	libssl.so.1.0.0 => /usr/local/lib/libssl.so.1.0.0 (0x00007fa767f85000)

I'm uncertain who is to blame. The nghttpd script or automake or.. me? My source distro times are long gone ;-)

**Comment by tatsuhiro-t** [comment]
This is because autotools does not use rpath or similar things.  The end result is ld always try to load standard location first.
Usually I use LD_LIBRARY_PATH to sort these things out.  But you can specify some linker flags and spare some typings.  For example, you can do:
```
./configure LDFLAGS="-Wl,-rpath,/usr/local/lib/"
```
to set rpath in binary to load library under /usr/local/lib.

**Comment by bekoli** [comment]
Thanks a lot, I learned something today :+1: 

Blaming autotools and setting rpath works fine for me.
